// @milkdown/plugin-history@7.2.3/lib/index.es.js
import { commandsCtx as e } from "@milkdown/core";
import { undo as y, redo as c, history as h } from "@milkdown/prose/history";
import { $command as i, $ctx as p, $prose as l, $useKeymap as u } from "@milkdown/utils";
const t = (o, r) => (Object.assign(o, {
  meta: {
    package: "@milkdown/plugin-history",
    ...r
  }
}), o), s = i("Undo", () => () => y);
t(s, {
  displayName: "Command<undo>"
});
const d = i("Undo", () => () => c);
t(s, {
  displayName: "Command<redo>"
});
const m = p({}, "historyProviderConfig");
t(m, {
  displayName: "Ctx<historyProviderConfig>"
});
const n = l((o) => h(o.get(m.key)));
t(n, {
  displayName: "Ctx<historyProviderPlugin>"
});
const a = u("historyKeymap", {
  Undo: {
    shortcuts: "Mod-z",
    command: (o) => {
      const r = o.get(e);
      return () => r.call(s.key);
    }
  },
  Redo: {
    shortcuts: ["Mod-y", "Shift-Mod-z"],
    command: (o) => {
      const r = o.get(e);
      return () => r.call(d.key);
    }
  }
});
t(a.ctx, {
  displayName: "KeymapCtx<history>"
});
t(a.shortcuts, {
  displayName: "Keymap<history>"
});
const x = [m, n, a, s, d].flat();
export {
  x as history,
  a as historyKeymap,
  m as historyProviderConfig,
  n as historyProviderPlugin,
  d as redoCommand,
  s as undoCommand
};
